content/en/BestPractices/prevent-xml-external-entities-attack.md

```diff
@@ -0,0 +1,292 @@
+  </company>
+```
+
+However, this could be prevented by using `setFeature` to disable the DOCTYPE declaration and remediate the XXE attack, as shown in the below code example: 
+
+```
+  SAXParserFactory factory = SAXParserFactory.newInstance();
+
+  try {
+
+      // https://rules.sonarsource.com/java/RSPEC-2755
+      // prevent XXE, completely disable DOCTYPE declaration:
+      factory.setFeature("http://apache.org/xml/features/disallow-doctype-decl", true);
+
+      SAXParser saxParser = factory.newSAXParser();
+
+      PrintAllHandlerSax handler = new PrintAllHandlerSax();
+
+      saxParser.parse(FILENAME, handler);
+
+  } catch (ParserConfigurationException | SAXException | IOException e) {
+      e.printStackTrace();
+  }
+```
+
+You may refer to different Sonar Source Rules to understand the XXE remediation for famous XML parsers using rules such as https://rules.sonarsource.com/java/RSPEC-2755.
+
+
+#### Denial of Service - XML Billion Laugh Attack 
+
+XML parsers are prone to denial of service attacks when a malicious XML document containing the same large entity is repeated repeatedly, causing an infinite processing loop. In addition, due to the lack of restrictions in place, such as a limit on the number of entity expansions, the XML processor can consume a heavy amount of memory and time during the parsing resulting in a Denial of Service attack. 
+
+For example, in the below code example, when the `XMLConstants.FEATURE_SECURE_PROCESSING` feature is set to `false`, it could allow an attacker to perform an XML Billion Laugh Attack: 
+
+```
+SAXReader xmlReader = new SAXReader();
+xmlReader.setFeature(XMLConstants.FEATURE_SECURE_PROCESSING, false);
+```
+An attacker can craft a billion laugh attack payload like the following to cause a denial of service in the application system: 
+
+```
+<?xml version=”1.0" encoding=”UTF-8"?>
+    
+    <!DOCTYPE example [
+    
+    <!ELEMENT example ANY >
+    
+    <!ENTITY lol “lol”>
+    
+    <!ENTITY lol1 “&lol;&lol;&lol;&lol;&lol;&lol;&lol;&lol;&lol;&lol;”>
+    
+    <!ENTITY lol2 “&lol1;&lol1;&lol1;&lol1;&lol1;&lol1;&lol1;&lol1;&lol1;&lol1;”>
+    
+    <!ENTITY lol3 “&lol2;&lol2;&lol2;&lol2;&lol2;&lol2;&lol2;&lol2;&lol2;&lol2;”>
+    
+    <!ENTITY lol4 “&lol3;&lol3;&lol3;&lol3;&lol3;&lol3;&lol3;&lol3;&lol3;&lol3;”>
+    
+    <!ENTITY lol5 “&lol4;&lol4;&lol4;&lol4;&lol4;&lol4;&lol4;&lol4;&lol4;&lol4;”>
+    
+    <!ENTITY lol6 “&lol5;&lol5;&lol5;&lol5;&lol5;&lol5;&lol5;&lol5;&lol5;&lol5;”>
+    
+    <!ENTITY lol7 “&lol6;&lol6;&lol6;&lol6;&lol6;&lol6;&lol6;&lol6;&lol6;&lol6;”>
+    
+    <!ENTITY lol8 “&lol7;&lol7;&lol7;&lol7;&lol7;&lol7;&lol7;&lol7;&lol7;&lol7;”>
+    
+    <!ENTITY lol9 “&lol8;&lol8;&lol8;&lol8;&lol8;&lol8;&lol8;&lol8;&lol8;&lol8;”>
+    
+    ]>
+    
+    <example>&lol9;</example>
+```
+
+However, to remediate this situation, further validation can be added via setting `XMLConstants.FEATURE_SECURE_PROCESSING` as `true` like in the code below:
+
+```
```

Comment:
```suggestion
To prevent this situation, add further validation by setting `XMLConstants.FEATURE_SECURE_PROCESSING` to `true`, as shown in the following example:

```xml
```